File: src/ports/driven/repositories/catalog-product.repository.ts
// src/ports/driven/repositories/catalog-product.repository.ts

import { Context, Effect, Option } from "effect"
import type { CatalogProduct } from "../../../domain/catalog"
import type { ProductId } from "../../../domain/pilot"

// ============================================
// CATALOG PRODUCT REPOSITORY (Read Model)
// ============================================

export interface CatalogProductRepository {
  readonly upsert: (
    product: CatalogProduct
  ) => Effect.Effect<CatalogProduct, unknown>

  readonly findById: (
    id: ProductId
  ) => Effect.Effect<Option.Option<CatalogProduct>, unknown>

  readonly findAll: () => Effect.Effect<readonly CatalogProduct[], unknown>
}

export const CatalogProductRepository = Context.GenericTag<CatalogProductRepository>(
  "CatalogProductRepository"
)

File: apps/server-java/src/main/java/com/maisonamane/port/driven/repository/CatalogProductRepository.java
package com.maisonamane.port.driven.repository;

import com.maisonamane.domain.catalog.projection.CatalogProduct;
import com.maisonamane.domain.pilot.valueobject.ProductId;
import com.maisonamane.port.driven.error.PersistenceError;
import io.vavr.control.Either;
import io.vavr.control.Option;
import java.util.List;

/**
 * Repository interface for CatalogProduct (Read Model).
 * Maps Effect-TS Context.Tag interface to Java interface with Either returns.
 */
public interface CatalogProductRepository {

    Either<PersistenceError, CatalogProduct> upsert(CatalogProduct product);

    Either<PersistenceError, Option<CatalogProduct>> findById(ProductId id);

    Either<PersistenceError, List<CatalogProduct>> findAll();
}
